### PR TITLE
Create more validation tests for PUT/PATCH logic

### LINKED ISSUE #47058: Create more validation tests for PUT/PATCH logic
### Description

As seen here: https://github.com/keycloak/keycloak/pull/46994/changes#diff-226b89fef493191617383e6a89fd93ca12854e2d5f5a52140be09e7a259c897aR143, we have some gap in our testsuite related to validation tests for PUT/PATCH. 

We should verify that all the validations from the Default group are applied when using the PUT/PATCH logic. 
Suggested test cases:
- Tests PUT with:
    - Invalid appUrl
    - Invalid redirectUri
    - Blank role
    - Blank webOrigin
- Tests PATCH with:
    - Invalid appUrl
    - Invalid redirectUri
    - Blank role
    - Blank webOrigin

### PR DESCRIPTION
**Pull request by mabartos**
- Closes #47058
- Number of tests increased from 19 to 99 :rocket: 
- There are many common tests that run automatically with POST/PUT/PATCH (bigger coverage)
- Tests run with OIDC and SAML (bigger coverage)
- Improving the error message for invalid type of field with PATCH
- When the `clientId` is specified in the PATCH payload, it needs to be aligned with the path client ID
- Added more tests for client-secret-based validations
- Covering more corner cases

Based on my analysis, we should cover all validation rules for now. 

@vmuzikar @edewit Could you please check it? Thanks!

### CODE REVIEW

**Review by edewit — APPROVED** [review]

**Comment by mabartos** [comment]
@vmuzikar Could you please check it? Thanks!

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+        return getPayloadClientId(true);
+    }
+
+    // ===========================================
+    // clientId validation tests
+    // ===========================================
+
+    @ParameterizedTest
+    @ValueSource(strings = {OIDCClientRepresentation.PROTOCOL, SAMLClientRepresentation.PROTOCOL})
+    public void clientWithoutClientIdFails(String protocol) throws Exception {
```
Comment (vmuzikar) [inline rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java:107]: We can mark the clientId field as required as mentioned by @sschu on the last sync. But we can do that as a follow-up.

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            assertThat(response.getStatusLine().getStatusCode(), is(400));
+
+            String responseBody = EntityUtils.toString(response.getEntity());
+            assertThat(responseBody, containsString("Cannot replace client resource with null"));
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {OIDCClientRepresentation.PROTOCOL, SAMLClientRepresentation.PROTOCOL})
+    public void clientWithTypeMismatchFails(String protocol) throws Exception {
```
Comment (vmuzikar) [inline rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java]: Is this a duplicate of `clientWithInvalidProtocolFails`? This test should rather validate that the protocol cannot be changed for existing clients.

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.hamcrest.CoreMatchers.hasItem;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+
+@KeycloakIntegrationTest(config = AbstractClientValidationTest.AdminV2Config.class)
+public class PatchClientValidationTest extends PutClientValidationTest {
```
Comment (vmuzikar) [inline rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java]: We should not be extending the put test here. It's a bit cumbersome then to exclude the non-relevant tests for patch. In the put test we're even hardcoding the `HttpPut` requests in some tests, effectively running them here again. Please rather move the few tests that should be shared between patch and put to the abstract class and explicitly exclude them for post. This approach is quite fragile.

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java
Comment (vmuzikar) [inline rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java:1]: A generic note. For tests that are creating clients, should we register them in the clean up logic so they are properly deleted later? I think it's important with the current push for fast test runs which can also include re-using running server instances.

**Review by vmuzikar — COMMENTED** [review]
@mabartos Thank you for the PR!

**Comment by vmuzikar** [comment]
This conflicts/overlaps with https://github.com/keycloak/keycloak/pull/47375. CC @edewit

**Comment by vmuzikar** [comment]
@Pepo48 @shawkins @michalvavrik Can you please review as well?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java [hunk 4]
```diff
@@ -232,8 +231,15 @@ public BaseClientRepresentation patchClient(RealmModel realm, String clientId, P
                         // based on the RFC 7396 JSON Merge Patch should replace the whole entity if the patch is not an object - we can't do it
                         throw new ServiceException("Cannot replace client resource with null", Response.Status.BAD_REQUEST);
                     }
+                    if (patch.has("clientId")) {
+                        // we don't allow renaming via patch yet
+                        assertSameClientIds(clientId, patch.get("clientId").asText());
```
Comment (shawkins) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java]: Why have this check in two places? Couldn't it just be moved to createOrUpdate?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java [hunk 5]
```diff
@@ -263,6 +269,12 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
         fireAdminEvent(OperationType.DELETE, client);
     }
 
+    protected void assertSameClientIds(String pathId, String payloadId) {
+        if (!Objects.equals(pathId, payloadId)) {
```
Comment (michalvavrik) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java:271]: If someone left the client id out of the patch, here `payloadId` would be "null" string. Maybe you could improve it to say something like "clientId cannot be removed via PATCH"?

**Review by michalvavrik — COMMENTED** [review]
I didn't find a test that tries to change protocol from OIDC to SAML. I hope I didn't miss something, can you double-check, please?

**Comment by michalvavrik** [comment]
Sorry, I didn't read existing comments first, it was already noted here https://github.com/keycloak/keycloak/pull/47282/#discussion_r2989637224 so forget about it.

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java (hunk 3 above)
Comment (mabartos) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java]: Ok, we can do it this way.

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java (hunk 2 above)
Comment (mabartos) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java]: There's a small overlap between them. But the `clientWithTypeMismatchFails` also test the `JsonMappingException` handling in the DefaultClientService#patchClient. 

Test updated: https://github.com/keycloak/keycloak/pull/47282/commits/799c845b60a80ee7ab2b2476ca70d0c908633574

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java (hunk 3 above)
Comment (mabartos) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/PatchClientValidationTest.java]: Done in: https://github.com/keycloak/keycloak/pull/47282/commits/6e7c7862f83b124b3e8e9c5ffc948d46c50d0367

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java
Comment (mabartos) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java:1]: @vmuzikar Should not be required. We're using the managed realm, managed clients, and manual clients. Managed realm has CLASS lifecycle, so everything attached to it should be torn down during the "AfterClass/AfterAll". The same with the managed clients, as they have also the CLASS lifecycle, and are attached to the managed realm. For manual clients, they're attached to the managed realm, so they should also be removed once the managed realm is about to be killed. For creating the manual clients, we just need to ensure they're unique for the test class. 

So, AFAIK, we should be fine :)

**Comment by mabartos** [comment]
> I didn't find a test that tries to change protocol from OIDC to SAML. I hope I didn't miss something, can you double-check, please?

@michalvavrik It's similar to this, and should be already fixed: https://github.com/keycloak/keycloak/pull/47282#discussion_r2989730449. I'll verify if we have it verified on the PUT as well.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java (hunk 4 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java]: Updated

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java (hunk 5 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java:271]: Maybe also related to the required clientId? https://github.com/keycloak/keycloak/pull/47282#discussion_r2989587815

As that might be done in a follow-up.

Comment (michalvavrik) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java:271]: Ok

**Review by michalvavrik — APPROVED** [review]

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java (hunk 1 above)
Comment (mabartos) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java:107]: Created follow-up: 
- https://github.com/keycloak/keycloak/issues/47524

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java (hunk 5 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/DefaultClientService.java:271]: Follow-up:
- https://github.com/keycloak/keycloak/issues/47524

**Comment by mabartos** [comment]
@vmuzikar Can you review it again? Thanks!

File: rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java
Comment (vmuzikar) [reply rest/admin-v2/tests/src/test/java/org/keycloak/tests/admin/client/v2/validation/AbstractClientValidationTest.java:1]: Ok, good point, let's keep as is.

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks for the updates.
